src/coreclr/vm/jithelpers.cpp

```diff
@@ -2331,6 +2087,17 @@ NOINLINE HCIMPL3(CORINFO_MethodPtr, JIT_VirtualFunctionPointer_Framed, Object *
     MethodDesc* pStaticMD = (MethodDesc*) methodHnd;
     TypeHandle staticTH(classHnd);
 
+    if (staticTH.IsNull())
+    {
+        // This may be NULL on input for cases where the methodHnd is not an interface method, or if getting the method table from the 
+        // MethodDesc will return an exact type.
+        if (pStaticMD->IsInterface())
+        {
+            staticTH = pStaticMD->GetMethodTable();
+            _ASSERTE(!staticTH.IsCanonicalSubtype());
+        }
+    }
+
     pStaticMD->CheckRestore();
 
     // MDIL: If IL specifies callvirt/ldvirtftn it remains a "virtual" instruction
```

Comment:
We're supposed to support this sort of adjustment as part of ReadyToRun (since we support converting a non-virtual method into a virtual method); however, I just checked, and we don't actually have testing for the case where we construct a delegate from the target method, and the behavior at runtime is incorrect in this scenario. (We do have tests for actual dispatch conversion into a virtual method.) I intend to update the comment here, but leave the logic in place, and file a bug about the R2R behavior.

Reply:
Sigh. Nevermind my previous analysis.

1. Converting from a non-virtual to a virtual in ReadyToRun is not a safe change from the point of view of ReadyToRun. See https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules for the rules that we actively support, so an affirmative test in C# should not be written.
2. However, we still need to ensure that code doesn't fail in surprising ways with ReadyToRun, so we still need this logic if a customer converts from Virtual to Non-Virtual, which is ALSO a breaking change, but we can hit it if a customer fails to follow all of the rules in the above document. Keeping the existing behavior is done so that failure modes of ReadyToRun generated code are still broadly similar to the failure modes of non ReadyToRun code.